In [1]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>


In [2]:
tests = [
    {
        "name": "Example 1",
        "starting_numbers": "0,3,6",
        "expected": 436,
    },
    {
        "name": "Example 2",
        "starting_numbers": "1,3,2",
        "expected": 1,
    },
    {
        "name": "Example 3",
        "starting_numbers": "2,1,3",
        "expected": 10,
    },
    {
        "name": "Example 4",
        "starting_numbers": "1,2,3",
        "expected": 27,
    },
    {
        "name": "Example 5",
        "starting_numbers": "2,3,1",
        "expected": 78,
    },
    {
        "name": "Example 6",
        "starting_numbers": "3,2,1",
        "expected": 438,
    },
    {
        "name": "Example 7",
        "starting_numbers": "3,1,2",
        "expected": 1836,
    },
]


def nth_number_spoken(starting_numbers: str, nth: int = 2020) -> int:
    starting_numbers = [int(num) for num in starting_numbers.split(",")]
    numbers_spoken = {}

    for i, num in enumerate(starting_numbers):
        numbers_spoken[num] = i + 1

    most_recently_spoken = 0

    for current_number in range(len(numbers_spoken) + 1, nth):
        if most_recently_spoken in numbers_spoken:
            diff = current_number - numbers_spoken[most_recently_spoken]
            numbers_spoken[most_recently_spoken] = current_number
        else:
            diff = 0
            numbers_spoken[most_recently_spoken] = current_number

        most_recently_spoken = diff

    return most_recently_spoken


# nth_number_spoken("0,3,6")
run_tests_params(nth_number_spoken, tests)


Test Example 1 passed, for nth_number_spoken.
Test Example 2 passed, for nth_number_spoken.
Test Example 3 passed, for nth_number_spoken.
Test Example 4 passed, for nth_number_spoken.
Test Example 5 passed, for nth_number_spoken.
Test Example 6 passed, for nth_number_spoken.
Test Example 7 passed, for nth_number_spoken.
Success


In [3]:
puzzle = "17,1,3,16,19,0"

print(f"Part I: {nth_number_spoken(puzzle)}")

Part I: 694


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>694</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>Impressed, the Elves issue you a challenge: determine the <code>30000000</code>th number spoken. For example, given the same starting numbers as above:</p>
<ul>
<li>Given <code>0,3,6</code>, the <code>30000000</code>th number spoken is <code>175594</code>.</li>
<li>Given <code>1,3,2</code>, the <code>30000000</code>th number spoken is <code>2578</code>.</li>
<li>Given <code>2,1,3</code>, the <code>30000000</code>th number spoken is <code>3544142</code>.</li>
<li>Given <code>1,2,3</code>, the <code>30000000</code>th number spoken is <code>261214</code>.</li>
<li>Given <code>2,3,1</code>, the <code>30000000</code>th number spoken is <code>6895259</code>.</li>
<li>Given <code>3,2,1</code>, the <code>30000000</code>th number spoken is <code>18</code>.</li>
<li>Given <code>3,1,2</code>, the <code>30000000</code>th number spoken is <code>362</code>.</li>
</ul>
<p>Given your starting numbers, <em>what will be the <code>30000000</code>th number spoken?</em></p>
</article>

</main>


In [4]:
from functools import partial
from timeit import default_timer as timer


tests = [
    {
        "name": "Example 1",
        "starting_numbers": "0,3,6",
        "expected": 175594,
    },
    {
        "name": "Example 2",
        "starting_numbers": "1,3,2",
        "expected": 2578,
    },
    {
        "name": "Example 3",
        "starting_numbers": "2,1,3",
        "expected": 3544142,
    },
    {
        "name": "Example 4",
        "starting_numbers": "1,2,3",
        "expected": 261214,
    },
    {
        "name": "Example 5",
        "starting_numbers": "2,3,1",
        "expected": 6895259,
    },
    {
        "name": "Example 6",
        "starting_numbers": "3,2,1",
        "expected": 18,
    },
    {
        "name": "Example 7",
        "starting_numbers": "3,1,2",
        "expected": 362,
    },
]


the_30000000th_number = partial(nth_number_spoken, nth=30000000)
the_30000000th_number.__name__ = "the_30000000th_number"

start = timer()
run_tests_params(the_30000000th_number, tests)
print(f"\nTime elapsed: {timer() - start:.4f}s")


Test Example 1 passed, for the_30000000th_number.
Test Example 2 passed, for the_30000000th_number.
Test Example 3 passed, for the_30000000th_number.
Test Example 4 passed, for the_30000000th_number.
Test Example 5 passed, for the_30000000th_number.
Test Example 6 passed, for the_30000000th_number.
Test Example 7 passed, for the_30000000th_number.
Success

Time elapsed: 70.2477s


> Don't Know (the Van Eck Sequence) - Numberphile <br/> > [Don't Know (the Van Eck Sequence) - Numberphile](https://youtu.be/etMJxB-igrc)


In [5]:
start = timer()
print(f"Part II: {the_30000000th_number(puzzle)}")
print(f"\nTime elapsed: {timer() - start:.4f}s")

Part II: 21768614

Time elapsed: 10.1433s


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>21768614</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>

</main>
